docs/data/data-grid/pagination/pagination.md

```diff
@@ -94,40 +94,155 @@ const [paginationModel, setPaginationModel] = React.useState({
+The index-based pagination is the most common type of pagination. It follows the same pattern as the client-side pagination (page, pageSize), but the data is fetched from the server.
 
-Since the `rowCount` prop is used to compute the number of available pages, switching it to `undefined` during loading resets the page to zero.
-To avoid this problem, you can keep the previous value of `rowCount` while loading as follows:
+To enable server-side pagination, you need to:
 
-```jsx
-const [rowCountState, setRowCountState] = React.useState(rowCount);
-React.useEffect(() => {
-  setRowCountState((prevRowCountState) =>
-    rowCount !== undefined ? rowCount : prevRowCountState,
-  );
-}, [rowCount, setRowCountState]);
-
-<DataGrid rowCount={rowCountState} />;
-```
+- Set the `paginationMode` prop to `server`
+- Use the `onPaginationModelChange` prop to react to the page changes and load the data from the server
+
+The server-side pagination can be further categorized into sub-types based on the availability of the total number of rows or `rowCount`.
+
+The Grid uses the `rowCount` to calculate the number of pages and to show the information about the current state of the pagination in the footer.
+You can provide the `rowCount` in one of the following ways:
+
+- Initialize - use `initialState.pagination.rowCount` prop to initialize the `rowCount`
+- Control - use the `rowCount` prop along with `onRowCountChange` to control the `rowCount` and reflect the changes when the row count is updated
+- Set using the `apiRef` - use the `apiRef.current.setRowCount` method to set the `rowCount` after the Grid is initialized
+
+There can be three different possibilities regarding the availability of the `rowCount` on the server-side:
+
+1. Row count is available (known)
+2. Row count is not available (unknown)
+3. Row count is available but is not accurate and may update later on (estimated)
 
 :::warning
 The `rowCount` prop is used in server-side pagination mode to inform the DataGrid about the total number of rows in your dataset.
 This prop is ignored when the `paginationMode` is set to `client`, i.e. when the pagination is handled on the client-side.
 :::
 
+You can achieve each of these use-cases by setting the values of `rowCount`, `paginationMeta.hasNextPage`, and `estimatedRowCount` props. The following table summarizes the use of these props in each of these scenarios.
+
+|                              | Known row count | Unknown row count | Estimated row count |
+| :--------------------------- | :-------------- | :---------------- | :------------------ |
+| `rowCount`                   | `number`        | `-1`              | `-1`                |
+| `paginationMeta.hasNextPage` | —               | `boolean`         | `boolean`           |
+| `estimatedRowCount`          | —               | —                 | `number`            |
+
+The following examples demonstrate each of these scenarios:
+
+#### Known row count
```

Comment:
Could we keep the pagination buttons active while loading, as it currently is on `master`?
https://mui.com/x/react-data-grid/pagination/#basic-implementation
This allows switching pages faster if necessary.

Reply:
Thanks for linking to that issue, I wasn't aware of it.

> I am trying to use the datagrid in a way where server side pagination using cursors provides us the ability to go to the next page and reliably get the next results.

From what I can observe in https://deploy-preview-12474--material-ui-x.netlify.app/x/react-data-grid/pagination/#cursor-based-pagination, even though the pagination buttons aren't disabled while loading, I can't skip pages by clicking while loading 👍🏻